src/models/assistant-model.ts

```diff
@@ -115,7 +115,27 @@ export const AssistantModel = types
   }))
   .actions((self) => ({
     addMessageToCODAPNotificationQueue(msg: string) {
-      self.codapNotificationQueue.push(msg);
+      const extracted = extractDataContexts(msg);
+      if (extracted) {
+        self.codapNotificationQueue.push(extracted);
+      }
+    },
+    clearCODAPNotificationQueue() {
+      self.codapNotificationQueue.clear();
+    },
+    clearUserMessageQueue() {
+      self.messageQueue.clear();
+    },
+    deDupeCODAPNotificationQueue(msg: IExtractedDataContext) {
+      // If any existing messages in the queue have the same type and context name values as `msg`, we remove them
+      // from the queue since they are no longer up to date and will be wasting valuable space.
+      self.codapNotificationQueue.replace(
+        self.codapNotificationQueue.filter(queued => 
+          !(queued.type === msg.type && 
+            Object.keys(queued.codapData)[0] === Object.keys(msg.codapData)[0] // context name check
```

Comment:
Are you sure that the `codapData` objects will have exactly one key each? Because if they have more than one it's not obvious that the same one will get returned by the `[0]`.